### PR TITLE
Add Function to retrieve all Keys in a Hash

### PR DESCRIPTION
**Pull request by dedmen**
I am currently doing
`params ["", "_keys"]` to get the keys of a hash. 
But that might break if the internal structure changes. 
So rather have a offical API function for that.

### CODE REVIEW

File: addons/hashes/CfgFunctions.hpp [hunk 1]
```diff
@@ -26,6 +26,12 @@ class CfgFunctions
                 description = "Iterate through all keys and values in a Hash. Code must return a bool, if false then key is removed";
                 file = "\x\cba\addons\hashes\fnc_hashFilter.sqf";
             };
+            // CBA_fnc_hashGetKeys
+            class hashGetKeys
+            {
+                description = "Returns all the Keys in a Hash";
```
Comment (commy2) [inline addons/hashes/CfgFunctions.hpp]: Returns all ~the~ Keys in a Hash _._

File: addons/hashes/fnc_hashGetKeys.sqf [hunk 2]
```diff
@@ -0,0 +1,27 @@
+/* ----------------------------------------------------------------------------
+Function: CBA_fnc_hashGetKeys
+
+Description:
+    Returns all the Keys in a Hash
```
Comment (commy2) [inline addons/hashes/fnc_hashGetKeys.sqf]: Same as above.

**Comment by dedmen** [comment]
My fantasy to make up tests is not that good.

**Comment by commy2** [comment]
test fails, because of this line:
`TEST_OP(_keys,isEqualTo,["123","124",125],"hashSize");`
Array uses additional commas. Suggest to either use `ARR_3()` macro or to use the `_expected` variable.

**Comment by commy2** [comment]
Also put a period after "Returns all the Keys in a Hash" while you're at it. Thanks.

**Comment by dedmen** [comment]
All the pointy things!

**Comment by commy2** [comment]
```sqf
execVM "\x\cba\addons\hashes\test_hashes.sqf"
```

```
[CBA] (hashes) Test OK: (CBA_fnc_hashCreate is defined) x\cba\addons\hashes\test_hashes.sqf:14
[CBA] (hashes) Test OK: (CBA_fnc_hashGet is defined) x\cba\addons\hashes\test_hashes.sqf:15
[CBA] (hashes) Test OK: (CBA_fnc_hashSet is defined) x\cba\addons\hashes\test_hashes.sqf:16
[CBA] (hashes) Test OK: (CBA_fnc_hashHasKey is defined) x\cba\addons\hashes\test_hashes.sqf:17
[CBA] (hashes) Test OK: (CBA_fnc_isHash is defined) x\cba\addons\hashes\test_hashes.sqf:18
[CBA] (hashes) Test OK: (CBA_fnc_hashSize is defined) x\cba\addons\hashes\test_hashes.sqf:19
[CBA] (hashes) Test OK: (CBA_fnc_hashGetKeys is defined) x\cba\addons\hashes\test_hashes.sqf:20
[CBA] (hashes) Test OK: (not ([[]] call CBA_fnc_isHash)) x\cba\addons\hashes\test_hashes.sqf:22
[CBA] (hashes) Test OK: (not ([_hash] call CBA_fnc_isHash)) x\cba\addons\hashes\test_hashes.sqf:24
[CBA] (hashes) Test OK: (not ([5] call CBA_fnc_isHash)) x\cba\addons\hashes\test_hashes.sqf:25
[CBA] (hashes) Test OK: (_hash is defined) x\cba\addons\hashes\test_hashes.sqf:29
[CBA] (hashes) Test OK: ([_hash] call CBA_fnc_isHash) x\cba\addons\hashes\test_hashes.sqf:30
[CBA] (hashes) Test OK: (not (_result)) x\cba\addons\hashes\test_hashes.sqf:33
[CBA] (hashes) Test OK: ([_hash] call CBA_fnc_isHash) x\cba\addons\hashes\test_hashes.sqf:36
[CBA] (hashes) Test OK: (_result == 12) x\cba\addons\hashes\test_hashes.sqf:39
[CBA] (hashes) Test OK: (_result) x\cba\addons\hashes\test_hashes.sqf:42
[CBA] (hashes) Test OK: (not (_result)) x\cba\addons\hashes\test_hashes.sqf:45
[CBA] (hashes) Test OK: (isNil "_result") x\cba\addons\hashes\test_hashes.sqf:51
[CBA] (hashes) Test OK: (isNil "_result") x\cba\addons\hashes\test_hashes.sqf:58
[CBA] (hashes) Test OK: (isNil "_result") x\cba\addons\hashes\test_hashes.sqf:62
[CBA] (hashes) Test OK: ([_hash] call CBA_fnc_isHash) x\cba\addons\hashes\test_hashes.sqf:66
[CBA] (hashes) Test OK: (_result is defined) x\cba\addons\hashes\test_hashes.sqf:69
[CBA] (hashes) Test OK: (_result == 99) x\cba\addons\hashes\test_hashes.sqf:70
[CBA] (hashes) Test OK: ([_hash] call CBA_fnc_isHash) x\cba\addons\hashes\test_hashes.sqf:74
[CBA] (hashes) Test OK: (_result is defined) x\cba\addons\hashes\test_hashes.sqf:77
[CBA] (hashes) Test OK: (_result == -8) x\cba\addons\hashes\test_hashes.sqf:78
[CBA] (hashes) Test OK: (_result is defined) x\cba\addons\hashes\test_hashes.sqf:81
[CBA] (hashes) Test OK: (_result == 0) x\cba\addons\hashes\test_hashes.sqf:82
[CBA] (hashes) Test OK: (_result is defined) x\cba\addons\hashes\test_hashes.sqf:86
[CBA] (hashes) Test OK: (_result == 1) x\cba\addons\hashes\test_hashes.sqf:87
[CBA] (hashes) Test OK: (isNil "_result") x\cba\addons\hashes\test_hashes.sqf:91
[CBA] (hashes) ERROR: Test FAIL x\cba\addons\hashes\test_hashes.sqf:129
[CBA] (hashes) Test OK: (_size == 0) x\cba\addons\hashes\test_hashes.sqf:96
[CBA] (hashes) Test OK: (_size == 1) x\cba\addons\hashes\test_hashes.sqf:101
[CBA] (hashes) Test OK: (_size == 3) x\cba\addons\hashes\test_hashes.sqf:107
[CBA] (hashes) Test OK: (_size == 2) x\cba\addons\hashes\test_hashes.sqf:112
[CBA] (hashes) Test OK: (_size == -1) x\cba\addons\hashes\test_hashes.sqf:117
[CBA] (hashes) Test OK: (_keys isEqualTo []) x\cba\addons\hashes\test_hashes.sqf:122
            (_keys isEqualTo ["123", "124", 125])
```

File: addons/hashes/test_hashes.sqf [hunk 3]
```diff
@@ -115,4 +116,16 @@ _hash = [5, [4], [1], 2];
...
+// Empty hash keys
+_hash = [] call CBA_fnc_hashCreate;
+_keys = [_hash] call CBA_fnc_hashGetKeys;
+TEST_OP(_keys,isEqualTo,[],"hashGetKeys");
+
+// Two elements keys with different types
+[_hash, "123", 1] call CBA_fnc_hashSet;
+[_hash, "124", 2] call CBA_fnc_hashSet;
+[_hash, 125, 3] call CBA_fnc_hashSet;
+_keys = [_hash] call CBA_fnc_hashSize;
```
Comment (commy2) [inline addons/hashes/test_hashes.sqf]: `CBA_fnc_hashGetKeys`

File: addons/hashes/test_hashes.sqf [hunk 4]
```diff
@@ -115,4 +116,16 @@ _hash = [5, [4], [1], 2];
...
+TEST_OP(_keys,isEqualTo,[ARR_3("123","124",125)],"hashSize");
```
Comment (commy2) [inline addons/hashes/test_hashes.sqf]: `"hashGetKeys"`

**Comment by commy2** [comment]
Should this function be called `CBA_fnc_hashKeys` instead of `CBA_fnc_hashGetKeys`?

**Comment by dedmen** [comment]
I thought about that. But getKeys is more clear that you get something IMO.

**Comment by dedmen** [comment]
Thats what you get for copy pasting stuff.

**Comment by commy2** [comment]
```
[CBA] (hashes) Test OK: (CBA_fnc_hashCreate is defined) x\cba\addons\hashes\test_hashes.sqf:14
[CBA] (hashes) Test OK: (CBA_fnc_hashGet is defined) x\cba\addons\hashes\test_hashes.sqf:15
[CBA] (hashes) Test OK: (CBA_fnc_hashSet is defined) x\cba\addons\hashes\test_hashes.sqf:16
[CBA] (hashes) Test OK: (CBA_fnc_hashHasKey is defined) x\cba\addons\hashes\test_hashes.sqf:17
[CBA] (hashes) Test OK: (CBA_fnc_isHash is defined) x\cba\addons\hashes\test_hashes.sqf:18
[CBA] (hashes) Test OK: (CBA_fnc_hashSize is defined) x\cba\addons\hashes\test_hashes.sqf:19
[CBA] (hashes) Test OK: (CBA_fnc_hashGetKeys is defined) x\cba\addons\hashes\test_hashes.sqf:20
[CBA] (hashes) Test OK: (not ([[]] call CBA_fnc_isHash)) x\cba\addons\hashes\test_hashes.sqf:22
[CBA] (hashes) Test OK: (not ([_hash] call CBA_fnc_isHash)) x\cba\addons\hashes\test_hashes.sqf:24
[CBA] (hashes) Test OK: (not ([5] call CBA_fnc_isHash)) x\cba\addons\hashes\test_hashes.sqf:25
[CBA] (hashes) Test OK: (_hash is defined) x\cba\addons\hashes\test_hashes.sqf:29
[CBA] (hashes) Test OK: ([_hash] call CBA_fnc_isHash) x\cba\addons\hashes\test_hashes.sqf:30
[CBA] (hashes) Test OK: (not (_result)) x\cba\addons\hashes\test_hashes.sqf:33
[CBA] (hashes) Test OK: ([_hash] call CBA_fnc_isHash) x\cba\addons\hashes\test_hashes.sqf:36
[CBA] (hashes) Test OK: (_result == 12) x\cba\addons\hashes\test_hashes.sqf:39
[CBA] (hashes) Test OK: (_result) x\cba\addons\hashes\test_hashes.sqf:42
[CBA] (hashes) Test OK: (not (_result)) x\cba\addons\hashes\test_hashes.sqf:45
[CBA] (hashes) Test OK: (isNil "_result") x\cba\addons\hashes\test_hashes.sqf:51
[CBA] (hashes) Test OK: (isNil "_result") x\cba\addons\hashes\test_hashes.sqf:58
[CBA] (hashes) Test OK: (isNil "_result") x\cba\addons\hashes\test_hashes.sqf:62
[CBA] (hashes) Test OK: ([_hash] call CBA_fnc_isHash) x\cba\addons\hashes\test_hashes.sqf:66
[CBA] (hashes) Test OK: (_result is defined) x\cba\addons\hashes\test_hashes.sqf:69
[CBA] (hashes) Test OK: (_result == 99) x\cba\addons\hashes\test_hashes.sqf:70
[CBA] (hashes) Test OK: ([_hash] call CBA_fnc_isHash) x\cba\addons\hashes\test_hashes.sqf:74
[CBA] (hashes) Test OK: (_result is defined) x\cba\addons\hashes\test_hashes.sqf:77
[CBA] (hashes) Test OK: (_result == -8) x\cba\addons\hashes\test_hashes.sqf:78
[CBA] (hashes) Test OK: (_result is defined) x\cba\addons\hashes\test_hashes.sqf:81
[CBA] (hashes) Test OK: (_result == 0) x\cba\addons\hashes\test_hashes.sqf:82
[CBA] (hashes) Test OK: (_result is defined) x\cba\addons\hashes\test_hashes.sqf:86
[CBA] (hashes) Test OK: (_result == 1) x\cba\addons\hashes\test_hashes.sqf:87
[CBA] (hashes) Test OK: (isNil "_result") x\cba\addons\hashes\test_hashes.sqf:91
[CBA] (hashes) Test OK: (_size == 0) x\cba\addons\hashes\test_hashes.sqf:96
[CBA] (hashes) Test OK: (_size == 1) x\cba\addons\hashes\test_hashes.sqf:101
[CBA] (hashes) Test OK: (_size == 3) x\cba\addons\hashes\test_hashes.sqf:107
[CBA] (hashes) Test OK: (_size == 2) x\cba\addons\hashes\test_hashes.sqf:112
[CBA] (hashes) Test OK: (_size == -1) x\cba\addons\hashes\test_hashes.sqf:117
[CBA] (hashes) Test OK: (_keys isEqualTo []) x\cba\addons\hashes\test_hashes.sqf:122
[CBA] (hashes) Test OK: (_keys isEqualTo ["123", "124", 125]) x\cba\addons\hashes\test_hashes.sqf:129
```

File: addons/hashes/fnc_hashGetKeys.sqf [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+#include "script_component.hpp"
+#include "script_hashes.hpp"
+
+SCRIPT(hashGetKeys);
+
+// -----------------------------------------------------------------------------
+params [["_hash", [[], []], [[]]]];
+
+(_hash select HASH_KEYS)
```
Comment (commy2) [inline addons/hashes/fnc_hashGetKeys.sqf]: newline @ EOF

Should this array be copied?

Comment (dedmen) [inline addons/hashes/fnc_hashGetKeys.sqf]: What?! I swear I have asked if  you think it should be copied. Apparently I thought of it but never did it. My brain is smoosh. Yeah.. Copy is probably better But there may also be a use for byref

**Comment by dedmen** [comment]
Should finally be all fine now.

**Comment by commy2** [comment]
Now you put two newlines at the end.
Why is this so painful? /O\

**Comment by commy2** [comment]
I'd prefer `hashKeys`. No need to name a getter only function `getX` and it would just lead to confusions with `hashGet` which is a different thing alltogehter.

**Comment by dedmen** [comment]
I'll never make a small CBA function again. It's too much hassle ^^

File: addons/hashes/test_hashes.sqf [hunk 6]
```diff
@@ -115,4 +116,16 @@ _hash = [5, [4], [1], 2];
...
+_hash = [] call CBA_fnc_hashCreate;
+_keys = [_hash] call CBA_fnc_hashKeys;
+TEST_OP(_keys,isEqualTo,[],"hashGetKeys");
+
+// Two elements keys with different types
+[_hash, "123", 1] call CBA_fnc_hashSet;
+[_hash, "124", 2] call CBA_fnc_hashSet;
+[_hash, 125, 3] call CBA_fnc_hashSet;
+_keys = [_hash] call CBA_fnc_hashKeys;
+TEST_OP(_keys,isEqualTo,[ARR_3("123","124",125)],"hashGetKeys");
```
Comment (commy2) [inline addons/hashes/test_hashes.sqf]: hashKeys

File: addons/hashes/test_hashes.sqf [hunk 7]
```diff
@@ -115,4 +116,16 @@ _hash = [5, [4], [1], 2];
 _size = [_hash] call CBA_fnc_hashSize;
 TEST_OP(_size,==,-1,"hashSize");
 
+// Empty hash keys
+_hash = [] call CBA_fnc_hashCreate;
+_keys = [_hash] call CBA_fnc_hashKeys;
+TEST_OP(_keys,isEqualTo,[],"hashGetKeys");
```
Comment (commy2) [inline addons/hashes/test_hashes.sqf]: hashKeys

**Comment by commy2** [comment]
>I'll never make a small CBA function again. It's too much hassle ^^

I could've fixed it all in one go if I had push access to your branch. Maybe a secondary PR next time...

**Comment by dedmen** [comment]
Edits from maintainers are allowed :x So you should have access

**Comment by commy2** [comment]
lol
Why didn't you tell me earlier.

**Pull request merged**